Rewrite docs for core concepts (#395)
kitbagjs/router: docs/core-concepts/router.md

# Router
The router is responsible for matching routes to urls and updating the browser history. It also provides a way to navigate to routes, access the current route, and provides some routing utilities. See [Router](/api/types/Router.md) type for more information.

> ## Creating a Router

Reviewer: is it worth repeating this from quick start?

Create a router using the `createRouter` utility.

```ts {8}
import { createRoute, createRouter } from '@kitbag/router'

const routes = [
  createRoute(...),
  createRoute(...),
] as const

const router = createRouter(routes)
```

> ## Installing

Reviewer: is it worth repeating this from quick start?

Author: I think the quick start is always going to be duplicate of other docs. This page will likely grow to have more information and feel less repetitive. But I think these docs aren't repeated from quick start, quick start repeats them from here. If that makes sense.

Once you have created a router you can install it into a Vue application using the `use` method.

```ts {8}
import { createRouter } from '@kitbag/router'
import { createApp } from 'vue'
import App from './App.vue'

const router = createRouter(...)
const app = createApp(App)

app.use(router)
app.mount('#app')
```

> ## Type Safety

Reviewer: is it worth repeating this from quick start?

Author: Lets discuss this in general on this comment https://github.com/kitbagjs/router/pull/395#discussion_r1899187620

Kitbag Router utilizes [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) to provide the internal types to match the actual router you're using.

```ts
declare module '@kitbag/router' {
  interface Register {
    router: typeof router
  }
}
```

## Router Options
There are several options you can pass to the router to customize its behavior. See the [RouterOptions](/api/types/RouterOptions.md) type for more information and a list of options.

```ts
const router = createRouter(routes, {

>   base: '/app',

Reviewer: thoughts on expanding this example?

```suggestion
  base: '/app',
  initialUrl: '/home',
  historyMode: 'memory',
  prefetch: { components: true, props: false },
  rejections: { authNeeded: ... },
```

Author: I don't feel strongly, but I thought linking to the api docs for this would mean it would always be up to date and one less piece we have to manually manage.

})
```

## Router Instance

> The router created by `createRouter` can be used directly in your application. You can also access the router instance using the [useRouter](/composables/useRouter.md) composable within your components.

Reviewer: as you might have picked up on by now, I'd prefer our core-concepts to be as concise as possible. I personally wouldn't even mention the fact that you _could_ import your router, let's just tell them about `useRouter` as a good pattern for accessing the installed router

Author: I agree on being concise but also complete. I don't want to have a core concept and a deep dive page about the same topic. 

Also agree that `useRouter` is the preferred way of getting the router instance. Which is why that's the example here. I didn't intent to imply you could or should export/import it. Just that the return value of `createRouter` IS the router instance.

Author: We're going to continue to work on this page. We both agree its needed but not great in its current state.

### Push
Navigates to a specific route and adds a new history entry.

```ts
router.push('home')
```

### Replace
Replaces the current history entry with a new one.

```ts
router.replace('home')
```

### Refresh
Forces the router to re-evaluate the current route.

```ts
router.refresh()
```

### Back
Navigates to the previous history entry.

```ts
router.back()
```

### Forward
Navigates to the next history entry.

```ts
router.forward()
```

### Go
Moves the current history entry to a specific point in the history stack.

```ts
router.go(1)
```

### Reject
Handles route rejection based on a specified rejection type. See [Rejections](/advanced-concepts/rejections) for more information.

```ts
router.reject('NotFound')
```

### Find
Creates a URL for a given route and params.

```ts
router.find('users', { id: 1 })
```

### Resolve
Creates a ResolvedRoute record for a given route name and params.

```ts
router.resolve('users', { id: 1 })
```

### IsExternal
Checks if a given URL is external to the router instance.

```ts
router.isExternal('https://google.com')
```

### Install
Installs the router into a Vue application instance.

```ts
router.install(app)
```

### Start

> Initializes the router based on the initial route. Automatically called when the router is installed. Calling this more than once has no effect.

Reviewer: ```suggestion
Idempotent function that initializes the router based on the initial route. Automatically called when the router is installed.
```

Author: Idempotent is one of those words that always makes me feel like people are trying to sound smart and the average person doesn't actually know what it means lol. Do you feel strongly about not just explaining how it works?

Reviewer: lol fair enough. It's a good word that devs should know but I like simple
